iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_fifo.inl

```diff
@@ -31,27 +31,69 @@ inline bool SpscFifo<ValueType, Capacity>::push(const ValueType& value) noexcept
-    m_write_pos.store(currentWritePos + 1, std::memory_order_release);
+    // SYNC POINT WRITE: m_data
+    // We need to make sure that writing the value happens before incrementing the
+    // m_writePos otherwise the following scenario can happen:
+    // 1. m_writePos is increased (but the value has not been written yet)
+    // 2. Another thread calls pop(): we check if the queue is empty => no
+    // 3. In pop(), when we read a value, a data race can occur when at the same time a value is
+    // written by push. With memory_order_release, this cannot happen as it is guaranteed that
+    // writing the data happens before incrementing m_writePos. Note that the following scenario
+    // can still happen (but this is not a problem):
+    // 1. A value is written (m_writePos hasn't been incremented yet)
+    // 2. Another thread calls pop(): we check if the queue is empty => yes
+    // 3. An element was already stored so we could have popped the element
+    m_writePos.store(currentWritePos + 1, std::memory_order_release);
     return true;
 }
 
 template <class ValueType, uint64_t Capacity>
 inline bool SpscFifo<ValueType, Capacity>::is_full() const noexcept
 {
-    return m_write_pos.load(std::memory_order_relaxed) == m_read_pos.load(std::memory_order_relaxed) + Capacity;
+    auto [readPos, writePos] = get_read_write_positions();
+    return writePos == readPos + Capacity;
 }
 
 template <class ValueType, uint64_t Capacity>
 inline uint64_t SpscFifo<ValueType, Capacity>::size() const noexcept
 {
-    return m_write_pos.load(std::memory_order_relaxed) - m_read_pos.load(std::memory_order_relaxed);
+    auto [readPos, writePos] = get_read_write_positions();
+    return writePos - readPos;
+}
+
+template <class ValueType, uint64_t Capacity>
+inline std::pair<uint64_t, uint64_t> SpscFifo<ValueType, Capacity>::get_read_write_positions() const noexcept
+{
+    uint64_t readPos{0};
+    uint64_t writePos{0};
+    do
+    {
+        readPos = m_readPos.load(std::memory_order_relaxed);
+        writePos = m_writePos.load(std::memory_order_relaxed);
+
+        // These checks are needed to make sure the result is consistent (but not necessarily
+        // up-to-date), otherwise the following scenario could happen (e.g. with the size() method):
+        // 1. We read m_writePos = 5
+        // 2. The OS puts the thread A to sleep
+        // 3. Other threads continue to call push and pop
+        // 4. New state m_writePos == 8, m_readPos == 6
+        // 5. The OS schedules again thread A and we read m_readPos = 6
+        // 6. We calculate 5 - 6 == -1  => overflow in uint64_t and size() returns
+        // 18446744073709551615
+        // 7. Another algorithm may depend on the reasonable argument that the size is always less or
+        // equal capacity it may lead to weird bugs.
+        // Note: it is still possible to return a size that is not up-to-date anymore but at least
+        // this size is logically valid
```

Comment:
I guess this is related to #2156 where we already concluded that this would be a misuse of the `SpscFifo`. I've the feeling you need a `MpmcFifo`. Do you have a use case for calling `size`, `empty` or `is_full` from another thread than the producer or consumer thread?

Note, you **cannot** rely on the `size` from any other thread than the producer or consumer thread
- the producer knows that the size will not increase but cannot tell whether is will be smaller right after the call returns
- the consumer knows that the size will not decrease but cannot tell whether it will be larger right after the call returns
- a third thread has absolutely no guarantees; it only knows that at some point in time it had this size but cannot tell whether it will be smaller or larger right after the call returns; the information is essentially useless; if you rely on this information you need to re-design your algorithm

Reply:
Ok, thanks for the explanation. It makes sense. I reverted the changes